### PR TITLE
Shared fine-tuned-adapter catalog handshake + NTK serving hybrid

### PR DESCRIPTION
**Pull request by alishalbaf**
## What

Two coordinated cogflow additions that let the platform serve ntkmirror-tuned LLMs natively (no LoRA conversion), reused by notebook callers too.

### `register_finetuned_catalog_entry` (+ async)
One catalog handshake for **any** fine-tuned adapter row (`lora` or `ntk_controller`). POSTs `/models/log` with the adapter `type`, the `base_model_id` self-FK, and the base's HuggingFace id. The catalog row's id is the tracking run id that already holds the artifact. Warn-only on POST failure (matches the LLM-catalog / `log_model` paths). `FINETUNED_ADAPTER_TYPES` is the single source of truth for the accepted vocabulary.

### Serving hybrid (`serve_llm` / `async_serve_llm` / `_build_llm_predictor`)
- `_build_llm_args` emits `--hf-overrides` (JSON, sorted keys) so the architecture can be redirected to a runtime-registered subclass (e.g. `NTKQwen2ForCausalLM`).
- New `hf_overrides` + `controller_storage_uri` params: HF base loaded via `--model_id` **plus** a side artifact (the controller) staged via `storageUri` to the predictor's `/mnt/models`. The KServe runtime's `get_model_id_or_path` lets `--model_id` win for the base while the storage-initializer stages the controller. S3 service account attached when the controller is S3-backed even on an HF base.

## Why
Consumed by Cog-Engine's `ntk_model` fine-tune + serving wiring (separate PR). Single helper so notebook and platform fine-tunes produce identical catalog rows.

## Tests
Sync + async catalog-entry coverage; `--hf-overrides` emission; HF-base-plus-controller hybrid spec. Full serving/models/async suites green, ruff clean.

## Follow-up (not in this PR)
Release as the next `release/v3.0.0bN` (PyPI + `cogflow:latest-beta`) so the Cog-Engine ISVC builder + kfp component pick it up.

### CODE REVIEW

File: cogflow/core/serving.py [hunk 1]
```diff
@@ -854,6 +867,19 @@ def _build_llm_predictor(
...
+            # artifact — e.g. an ntkmirror controller — is staged by
+            # KServe's storage-initializer to the predictor's model
+            # mount (``/mnt/models``). The runtime ignores the mount for
+            # the base (``--model_id`` wins in the HF server's
+            # ``get_model_id_or_path``), and the adapter-aware model
+            # subclass reads its artifact from there. Only valid in
+            # RawDeployment mode (the LLM default) — the storage-
+            # initializer's fieldRef env is rejected by the Knative
+            # webhook in Serverless mode.
+            model_block["storageUri"] = controller_storage_uri
```
[the remark this answers is not in the document]
Comment (alishalbaf) [reply cogflow/core/serving.py:912]: Fixed in 38da531 — `_build_llm_predictor` now raises `CogflowValidationError` when `controller_storage_uri` is set with a non-hf:// base. The single storageUri slot is claimed by an s3 base, so the hybrid only works with an HF base (loaded via --model_id).

File: cogflow/core/serving.py [hunk 2]
```diff
@@ -942,6 +974,13 @@ def deploy_llm(
         """Create a KServe InferenceService backed by the built-in
         ``huggingface`` ClusterServingRuntime (KServe 0.15+).
 
+        ``hf_overrides`` patches the loaded model's HF config (emitted as
+        the vLLM ``--hf-overrides`` flag) — used to redirect the
+        architecture to a runtime-registered adapter-aware subclass.
+        ``controller_storage_uri``, when set alongside an ``hf://`` base,
+        stages a side artifact (e.g. an ntkmirror controller) to the
+        predictor's model mount while the base still loads from the Hub.
+
```
[the remark this answers is not in the document]
Comment (alishalbaf) [reply cogflow/core/serving.py:1039]: Fixed in 38da531 — sync `serve_llm` now accepts and forwards `hf_overrides` + `controller_storage_uri` to `deploy_llm`, matching the async surface and the deploy_llm signature.

File: cogflow/core/serving.py [hunk 3]
```diff
@@ -772,6 +774,14 @@ def _build_llm_args(
...
         if not _is_unset_sentinel(normalized_kv_cache_dtype):
             args.append(f"--kv-cache-dtype={normalized_kv_cache_dtype}")
+        # ``--hf-overrides`` is a vLLM hyphen-style flag taking a JSON
+        # object that patches the loaded model's HF config. The platform
+        # uses it to redirect the architecture to a runtime-registered
+        # subclass (e.g. an adapter-aware ``NTK<Arch>ForCausalLM``)
+        # without re-uploading the base weights. ``sort_keys`` keeps the
+        # emitted JSON byte-stable so reruns produce a diff-clean ISVC.
+        if hf_overrides:
+            args.append(f"--hf-overrides={json.dumps(hf_overrides, sort_keys=True)}")
```
[the remark this answers is not in the document]
Comment (alishalbaf) [reply cogflow/core/serving.py]: Fixed in 3e8ff9f — hf_overrides is now validated as a JSON object (non-dict or non-serializable raises CogflowValidationError up-front).

File: cogflow/core/models.py [hunk 4]
```diff
@@ -1044,6 +1044,251 @@ async def async_register_llm_catalog_entry(
...
+            # sentinel (see the LLM-catalog and classical log paths).
+            "model_version": 0,
+            "register_date": datetime.fromtimestamp(register_date_ms / 1000).isoformat(),
+            "type": adapter_type,
+            "description": description,
+            "user_id": resolved_user,
+            # The self-FK to the base. The catalog service requires this
+            # for adapter rows (an adapter is meaningless without the
+            # base it modifies) and rejects the row otherwise.
+            "base_model_id": base_model_id,
```
[the remark this answers is not in the document]
Comment (alishalbaf) [reply cogflow/core/models.py]: Fixed in 3e8ff9f — base_model_id is now normalized via common.normalize_uuid in the catalog payload, consistent with model_id.

File: cogflow/core/serving.py [hunk 5]
```diff
@@ -772,6 +774,25 @@ def _build_llm_args(
...
+        if hf_overrides:
+            if not isinstance(hf_overrides, dict):
+                raise CogflowValidationError(
+                    f"hf_overrides must be a JSON object (dict), got {type(hf_overrides).__name__}"
+                )
+            try:
+                rendered = json.dumps(hf_overrides, sort_keys=True)
+            except (TypeError, ValueError) as exc:
+                raise CogflowValidationError(f"hf_overrides must be JSON-serializable: {exc}") from exc
+            args.append(f"--hf-overrides={rendered}")
```
[the remark this answers is not in the document]
Comment (alishalbaf) [reply cogflow/core/serving.py]: Fixed in ceaaac2 — the guard is now `hf_overrides is not None`, so falsey-but-invalid values ([], "", 0) are type-checked and rejected; an empty dict is accepted without emitting a no-op flag.

File: cogflow/core/serving.py [hunk 6]
```diff
@@ -996,6 +1065,8 @@ def deploy_llm(
             hf_secret_name=hf_secret_name,
             quantization=quantization,
             kv_cache_dtype=kv_cache_dtype,
+            hf_overrides=hf_overrides,
+            controller_storage_uri=controller_storage_uri,
         )
         predictor_spec, effective_annotations = ServingManager._apply_raw_deployment_defaults(
             predictor_spec, annotations
```
[the remark this answers is not in the document]
Comment (alishalbaf) [reply cogflow/core/serving.py:1098]: Fixed in cc2dcb4 — _assert_controller_compatible_mode now raises CogflowValidationError when controller_storage_uri is used with a non-RawDeployment effective mode (both sync + async deploy_llm), so a caller-forced Serverless deploymentMode fails fast instead of reconcile-failing under Knative.
